### PR TITLE
Replace `get_selector_from_name` and `get_storage_var_address` implementations

### LINKED ISSUE #580: Replace `get_selector_from_name` and `get_storage_var_address` implementations
### Feature Request

Write our own implementations of these.

### PR DESCRIPTION
**Pull request by drknzz**
Closes #580 

## Introduced changes

Remove uses of `starkware.starknet.public.api`
- replace `get_selector_from_name`
- replace `get_storage_var_address`

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/common.py [hunk 1]
```diff
@@ -40,3 +66,17 @@ def int_from_bytes(
     Converts the given bytes object (parsed according to the given byte order) to an integer.
     """
     return int.from_bytes(value, byteorder=byte_order, signed=signed)
+
+
+def get_storage_var_address(var_name: str, *args) -> int:
```
Comment (cptartur) [inline starknet_py/common.py:71]: Do we actually have any usecases of `args` in starknet.py? I know that cairo-lang's implementation had it but from my experience, I've never used it.

**Review by cptartur — COMMENTED** [review]
Two questions:

1. Should these method still be part of the public api? Perhaps it'd make sense to have underscores in their names.
2. Is eth-hash a good choice of a library? I know that cairo-lang uses it but [its readme](https://github.com/ethereum/eth-hash) recommends using higher level `eth_utils.keccak()`. Perhaps we should consider it?

File: starknet_py/common.py (hunk 1 above)
Comment (drknzz) [reply starknet_py/common.py:71]: As do we use `*args` in the code? We've got it in some places.

**Comment by drknzz** [comment]
> 1. Should these method still be part of the public api? Perhaps it'd make sense to have underscores in their names.

Well, for instance, the `Call` class is in public api. It takes `selector` argument, which is easier to pass as `get_selector_from_name("some_function")` and one can make sure it is the actual function they want to call by just looking at their code (instead of having an int as selector there).

> 2. Is eth-hash a good choice of a library? I know that cairo-lang uses it but [its readme](https://github.com/ethereum/eth-hash) recommends using higher level `eth_utils.keccak()`. Perhaps we should consider it?

You're right. Added the higher level `eth-utils` as a dependency.

File: starknet_py/common.py [hunk 2]
```diff
@@ -31,6 +40,23 @@ def int_from_hex(number: Union[str, int]) -> int:
...
+        return DEFAULT_ENTRY_POINT_SELECTOR
+
+    return starknet_keccak(data=func_name.encode("ascii"))
+
+
+def starknet_keccak(data: bytes) -> int:
+    """
+    A variant of eth-keccak that computes a value that fits in a StarkNet field element.
+    """
+    return int_from_bytes(keccak(data)) & MASK_250
```
Comment (cptartur) [inline starknet_py/common.py:57]: Perhaps at least this one should be kept private?

**Review by cptartur — COMMENTED** [review]
1. I though about this and I think this methods are core enough, that they should be kept public. But I added a comment to one I think shouldn't.
2. Good. I started wondering though, should we be concerned with the increased package size because of that? I think it's negligible, but perhaps it's not.

cc @Solpatium @war-in

File: starknet_py/common.py (hunk 1 above)
Comment (cptartur) [reply starknet_py/common.py:71]: I though we didn't. Let's keep it as is then.

File: starknet_py/common.py (hunk 2 above)
Comment (drknzz) [reply starknet_py/common.py:57]: Definitely could.

File: starknet_py/common.py [hunk 3]
```diff
@@ -31,6 +40,23 @@ def int_from_hex(number: Union[str, int]) -> int:
     return number if isinstance(number, int) else int(number, 16)
 
 
+def get_selector_from_name(func_name: str) -> int:
```
Comment (Solpatium) [inline starknet_py/common.py:43]: Can we move all those functions to a better place?

File: starknet_py/common.py [hunk 4]
```diff
@@ -40,3 +66,17 @@ def int_from_bytes(
...
+
+def get_storage_var_address(var_name: str, *args) -> int:
+    """
+    Returns the storage address of a StarkNet storage variable given its name and arguments.
+    """
+    res = starknet_keccak(var_name.encode("ascii"))
+
+    for arg in args:
+        if not isinstance(arg, int):
+            raise TypeError(f"Expected arguments to be integers. Found: {arg}.")
```
Comment (Solpatium) [inline starknet_py/common.py:79]: I think we can just type args properly.

**Review by Solpatium — CHANGES_REQUESTED** [review]

File: starknet_py/common.py (hunk 4 above)
Comment (drknzz) [reply starknet_py/common.py:79]: What exactly do you mean by typing args properly?

Comment (Solpatium) [reply starknet_py/common.py:79]: This: https://stackoverflow.com/a/37032111. Currently `get_storage_var_address` accepts anything, so it has to check types in the runtime.

File: starknet_py/common.py (hunk 3 above)
Comment (drknzz) [reply starknet_py/common.py:43]: Not sure what would be the best place for this.
`starknet_py.net.client_utils` maybe?
Or a new separate file?

Comment (Solpatium) [reply starknet_py/common.py:44]: I think cairo/selector makes sense. Selector is a part of the cairo program.

Comment (drknzz) [reply starknet_py/common.py:44]: What about `get_storage_var_address`?

Comment (Solpatium) [reply starknet_py/common.py:44]: `cairo/storage.py`? I like grouping things properly. @war-in how do you plan to arrange something like this in changed structure?

**Review by war-in — APPROVED** [review]
Great ✊ It can be merged after resolving @Solpatium's comment

File: starknet_py/constants.py [hunk 5]
```diff
@@ -16,4 +16,14 @@
...
 RPC_CLASS_HASH_NOT_FOUND_ERROR = 28
 
+DEFAULT_ENTRY_POINT_NAME = "__default__"
+DEFAULT_L1_ENTRY_POINT_NAME = "__l1_default__"
+DEFAULT_ENTRY_POINT_SELECTOR = 0
+MASK_250 = 2**250 - 1
+
+# MAX_STORAGE_ITEM_SIZE and ADDR_BOUND must be consistent with the corresponding constant in
+# starkware/starknet/common/storage/storage.cairo.
+MAX_STORAGE_ITEM_SIZE = 256
```
Comment (cptartur) [inline starknet_py/constants.py:26]: How likely is that this value will get increased in the future causing bugs in starknet.py? Is there any way to test it except for manual verification (probably not)?

File: starknet_py/common.py [hunk 6]
```diff
@@ -41,23 +31,6 @@ def int_from_hex(number: Union[str, int]) -> int:
     return number if isinstance(number, int) else int(number, 16)
```
Comment (Solpatium) [inline starknet_py/common.py:33]: Don't we want maintain backward compatibility with warnings?

**Review by Solpatium — CHANGES_REQUESTED** [review]

File: starknet_py/common.py (hunk 6 above)
Comment (drknzz) [reply starknet_py/common.py:33]: Where exactly?

File: starknet_py/constants.py (hunk 5 above)
Comment (Solpatium) [reply starknet_py/constants.py:26]: It might be changed in long future, but it will be a breaking change in starknet.

**Review by Solpatium — CHANGES_REQUESTED** [review]
Please add unit tests for added methods. I recommend generating some values using cairo-lang and making sure our functions return the same result.

**Pull request merged**